pkg/shed/index.go

```diff
@@ -322,102 +341,66 @@ func (f Index) Iterate(fn IndexIterFunc, options *IterateOptions) (err error) {
 			return err
 		}
 	}
-	it := f.db.NewIterator()
-	defer it.Release()
-
-	// move the cursor to the start key
-	ok := it.Seek(startKey)
-	if !ok {
-		f.logger.Debugf("seek error in Iterate. Error: %s", it.Error())
-		// stop iterator if seek has failed
-		return it.Error()
-	}
-	if options.SkipStartFromItem && bytes.Equal(startKey, it.Key()) {
-		// skip the start from Item if it is the first key
-		// and it is explicitly configured to skip it
-		ok = it.Next()
-	}
-	for ; ok; ok = it.Next() {
-		item, err := f.itemFromIterator(it, prefix)
+
+	err = f.db.Iterate(startKey, options.SkipStartFromItem, func(key []byte, value []byte) (stop bool, err error) {
+		item, err := f.itemFromKeyValue(key, value, prefix)
 		if err != nil {
-			if err == leveldb.ErrNotFound {
-				break
+			if err == badger.ErrKeyNotFound {
+				return true, nil
 			}
-			return err
-		}
-		stop, err := fn(item)
-		if err != nil {
-			f.logger.Debugf("error executing callback function in Iterate. Error: %s", err.Error())
-			return err
-		}
-		if stop {
-			break
+			return false, err
 		}
-	}
-	return it.Error()
+		stop, err = fn(item)
+		return stop, err
+	})
+	return err
 }
 
 // First returns the first item in the Index which encoded key starts with a prefix.
 // If the prefix is nil, the first element of the whole index is returned.
-// If Index has no elements, a leveldb.ErrNotFound error is returned.
+// If Index has no elements, a ErrNotFound error is returned.
 func (f Index) First(prefix []byte) (i Item, err error) {
-	it := f.db.NewIterator()
-	defer it.Release()
-
 	totalPrefix := append(f.prefix, prefix...)
-	it.Seek(totalPrefix)
-
-	return f.itemFromIterator(it, totalPrefix)
+	k, v, err := f.db.First(totalPrefix)
+	if err != nil {
+		return i, err
+	}
+	return f.itemFromKeyValue(k, v, totalPrefix)
 }
 
-// itemFromIterator returns the Item from the current iterator position.
+// itemFromKeyValue returns the Item from the current iterator position.
 // If the complete encoded key does not start with totalPrefix,
-// leveldb.ErrNotFound is returned. Value for totalPrefix must start with
+// ErrNotFound is returned. Value for totalPrefix must start with
```

Comment:
This comment change is not correct. It will return `badger.ErrKeyNotFound` not `ErrNotFound`.